### DISCUSSION TITLE
Not able to build bare metal Application SW targeting DDR

### DISCUSSION THREAD

**Discussion by geofflloyd** [polarfire-soc/polarfire-soc-documentation#122] [Bare metal embedded software] [answered]
[Booting bare metal Application SW to external DDR to run.pdf](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9067629/Booting.bare.metal.Application.SW.to.external.DDR.to.run.pdf)

I need help regarding booting our Application SW to external DDR to run. We are currently only running it from the Soft Console debugger out of the LIM and that works fine.

To date, I have done the following:
1>	Started with “mpfs-hal-ddr-demo” and “mpfs-hal-run-from-ddr-u54-1” sample projects
2>	Wrote Bootloader SW that runs on E51 only, boots and runs out of eNVM
3>	Bootloader SW mounts USB Thumbdrive FAT32 filesystem and PolarFire Dev Brd eMMC FAT32 filesystem both via FatFs.
4>	Bootloader SW provides menu to let you copy Application SW Image (*.bin) file from USB Thumbdrive to eMMC FAT32 filesystem
5>	Bootloader SW boots SW Image by reading file and writing it to DDR starting at base address of 0x8000_0000
6>	Bootloader SW then jumps to start of DDR at 0x8000_0000 using same approach found in “mpfs-hal-ddr-demo”

I need help with targeting our Application SW for external DDR and getting it to build to a *.bin file such that we can boot to it using the new Bootloader SW.
I am using the stock “mpfs-ddr-loaded-by-boot-loader.ld” and getting a lot of linker errors.
Need help modifying the linker file or using a different one to get the Application SW to build successfully.
In mss_sw_config.h file, have tried it both ways to no avail:
  => image loaded by bootloader (both 0 and 1)
  => hal clear memory (both 1 and 0)

Application SW runs fine out of LIM when booting to E51 and then starting U54-1 for main application.
Application SW also runs fine out of LIM with booting directly to U54-1 for main application.
But cannot get either to build successfully when targeting DDR.

Thanks, Geoff...

**Comment by hughbreslin** [comment]
Hi @geofflloyd 

can I check what version of the MPFS HAL is present in your bare metal project? You can find this in the [mpfs_hal_version](https://github.com/polarfire-soc/platform/blob/main/mpfs_hal/mpfs_hal_version.h) file.

One other thing to check, are your errors consistent when you change the settings in the `mss_sw_config.h` file? If so could you double check you're definitely modifying the correct version of this file based on the project include settings, e.g:
![image](https://user-images.githubusercontent.com/44862881/177979045-68909286-eb30-44b1-a829-e724a98f7aed.png)


One thing to be aware of with the `IMAGE_LOADED_BY_BOOTLOADER` flag is that this is intended for payloads running on a single core, its outlined in this [discussion](https://github.com/orgs/polarfire-soc/discussions/38) and we'll be releasing an update to support multiple harts. 

I'm not sure if this will be a solution but one thing you could try is following the steps in a similar way to the answer in this [discussion](https://github.com/polarfire-soc/polarfire-soc-documentation/discussions/109#discussioncomment-3056840) where instead of re-targeting the DDR linker to target the lim, re-target your working lim linker to target the DDR.

**Comment by geofflloyd** [comment]
[mpfs-lim.ld.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9072808/mpfs-lim.ld.txt)
[mss_sw_config.h.txt](https://github.com/polarfire-soc/polarfire-soc-documentation/files/9072809/mss_sw_config.h.txt)

My include path ordering used to be:
![Picture1](https://user-images.githubusercontent.com/94083386/178013308-a5d6cdaf-6077-4e65-8129-d34d03dcbe9d.png)

I changed the ordering per your suggestion:
![Picture2](https://user-images.githubusercontent.com/94083386/178013380-08920f49-c4ac-4ace-b091-62df4d24e4c0.png)
Still builds fine so will leave project using this new more logical ordering.

Microchip MPFS HAL Version:  (1.8.134)  Major.Minor.Patch

Both the mss_sw_confi.h and mpfs-lim.ld files that I am using to build successfully are attached as text files. Their locations in the project are listed below.

fledgling_app_sw\src\boards\icicle-kit-es\platform_config\mpfs_hal_config\mss_sw_config.h
    Changes to this file definitely effect the build to the compiler is using the file.

fledgling_app_sw\src\platform\platform_config_reference\linker\mpfs-lim.ld

Memory table from LIM linker file:
    envm (rx) : ORIGIN  = 0x20220100, LENGTH = 128k - 0x100
    dtim (rwx) : ORIGIN = 0x01000000, LENGTH = 7k
    switch_code (rx) : ORIGIN = 0x01001c00, LENGTH = 1k
    e51_itim (rwx) : ORIGIN = 0x01800000, LENGTH = 28k
    u54_1_itim (rwx) : ORIGIN = 0x01808000, LENGTH = 28k
    u54_2_itim (rwx) : ORIGIN = 0x01810000, LENGTH = 28k
    u54_3_itim (rwx) : ORIGIN = 0x01818000, LENGTH = 28k
    u54_4_itim (rwx) : ORIGIN = 0x01820000, LENGTH = 28k
    l2lim (rwx)      : ORIGIN = 0x08000000, LENGTH = 1280k
    scratchpad(rwx)  : ORIGIN = 0x0A000000, LENGTH = 0k
    /* DDR sections example */
    ddr_cached_32bit (rwx) : ORIGIN  = 0x80000000, LENGTH = 768M
    ddr_non_cached_32bit (rwx) : ORIGIN  = 0xC0000000, LENGTH = 256M
    ddr_wcb_32bit (rwx) : ORIGIN  = 0xD0000000, LENGTH = 256M
    ddr_cached_38bit (rwx) : ORIGIN  = 0x1000000000, LENGTH = 1024M
    ddr_non_cached_38bit (rwx) : ORIGIN  = 0x1400000000, LENGTH = 0k
    ddr_wcb_38bit (rwx) : ORIGIN  = 0x1800000000, LENGTH  = 0k

    HEAP_SIZE           = 8k;        /* needs to be calculated for your application if using */

    SIZE_OF_COMMON_HART_MEM = 4k;

    STACK_SIZE_E51_APPLICATION = 8k;
    STACK_SIZE_U54_1_APPLICATION = 32k;
    STACK_SIZE_U54_2_APPLICATION = 8k;
    STACK_SIZE_U54_3_APPLICATION = 8k;
    STACK_SIZE_U54_4_APPLICATION = 8k;
Note: I modified the MSS Configurator configuration to decrease scratchpad to 0K and increase l2lim to 1280K so that it matches above table settings.
Note: I also increased the U54_1 stack size from 8K to 32K.


Below are 3 iterations of same successful Application SW build targeting the LIM to show the region totals using 3 different Print Size settings. Useful to see totals, as well as details in both decimal and hex. Is there any way to show all 3 during 1 build?
=========================================================
Invoking: GNU RISC-V Cross Print Size
riscv64-unknown-elf-size --format=berkeley --totals --radix=10 "fledgling_app_sw.elf"
   text	   data	    bss	    dec	    hex	filename
 272700	   4624	 123472	 400796	  61d9c	fledgling_app_sw.elf
 272700	   4624	 123472	 400796	  61d9c	(TOTALS)
Finished building: fledgling_app_sw.siz
=========================================================
Invoking: GNU RISC-V Cross Print Size
riscv64-unknown-elf-size --format=sysv --totals --radix=10 "fledgling_app_sw.elf"
fledgling_app_sw.elf  :
section                size        addr
.text                272592   134217728
.l2_scratchpad            0   167772160
.ram_code               108    16784384
.sdata                  144   134490432
.data                  4480   134490576
.sbss                   432   134495056
.bss                  45216   134495488
.heap                  8192   134540704
.stack                65536   134549504
.app_hart_common       4096   134615040
.riscv.attributes        53           0
.comment                 81           0
.debug_line          433949           0
.debug_info         7339190           0
.debug_abbrev         52677           0
.debug_aranges        24480           0
.debug_str           465317           0
.debug_ranges         23904           0
.debug_macro         117168           0
.debug_frame          76448           0
Total               8934063
=========================================================
Invoking: GNU RISC-V Cross Print Size
riscv64-unknown-elf-size --format=sysv --totals --radix=16 "fledgling_app_sw.elf"
fledgling_app_sw.elf  :
section                 size        addr
.text                0x428d0   0x8000000
.l2_scratchpad           0x0   0xa000000
.ram_code               0x6c   0x1001c00
.sdata                  0x90   0x8042940
.data                 0x1180   0x80429d0
.sbss                  0x1b0   0x8043b50
.bss                  0xb0a0   0x8043d00
.heap                 0x2000   0x804eda0
.stack               0x10000   0x8051000
.app_hart_common      0x1000   0x8061000
.riscv.attributes       0x35         0x0
.comment                0x51         0x0
.debug_line          0x69f1d         0x0
.debug_info         0x6ffcb6         0x0
.debug_abbrev         0xcdc5         0x0
.debug_aranges        0x5fa0         0x0
.debug_str           0x719a5         0x0
.debug_ranges         0x5d60         0x0
.debug_macro         0x1c9b0         0x0
.debug_frame         0x12aa0         0x0
Total               0x8852af
=========================================================

Next, will try your suggestion of starting with the LIM linker file and setting the LIM sections to DDR sections and let you know what I find.

Some questions / assumptions:
1> Will start with Application SW that only targets U54-1 based on your feedback.
2> In mss_sw_config.h file, will set:
    => image loaded by bootloader to 1
    => hal clear memory to 0
3> Since my bootloader is blindly writing entire *.bin file to DDR starting at 0x8000_0000, then can't be any sections in the *.bin file that target the LIM or other memory regions. All linked code needs to get placed in the DDR regions.
4> Assume Bootloader SW must handle the startup code including setting up the DDR since Application SW starts running from the DDR.

Thanks, Geoff...
glloyd@srcinc.com
315-350-9197

**Comment by geofflloyd** [comment]
I just tried starting with the stock "mpfs-ddr-loaded-by-boot-loader.ld" linker file and changing the following:
    l2lim from 256K to 1280K to match my modified lim linker file.
    scratchpad from 256K to 0K to match my modified lim linker file.
    Heap Size from 0K to 8K to match my modified lim linker file.
    Stack Size U54 from 8K to 32K to match my modified lim linker file.
Still get the same linker errors. They are pasted below for reference.
Can you explain what is going wrong?

./src/application/filesystem.o ./src/application/fw.o ./src/application/i2c_interface.o ./src/application/menu_main.o ./src/application/menu_rf_test.o ./src/application/modulation_out.o ./src/application/mpfs.o ./src/application/ring_buffer.o ./src/application/serial_console.o ./src/application/serial_port.o ./src/application/spi_interface.o ./src/application/systick.o ./src/application/usb_host_msc.o   
./src/platform/mpfs_hal/startup_gcc/newlib_stubs.o: in function `_sbrk':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/newlib_stubs.c:278:(.text._sbrk+0xc): relocation truncated to fit: R_RISCV_HI20 against symbol `__heap_end' defined in .heap section in fledgling_app_sw.elf
./src/platform/mpfs_hal/startup_gcc/system_startup.o: in function `u54_single_hart':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/system_startup.c:230:(.text.u54_single_hart+0x2c): relocation truncated to fit: R_RISCV_HI20 against `.L7'
./src/platform/mpfs_hal/startup_gcc/system_startup.o: in function `init_memory':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/system_startup.c:487:(.text.init_memory+0x8): relocation truncated to fit: R_RISCV_HI20 against symbol `__text_end' defined in .text section in fledgling_app_sw.elf
./src/platform/mpfs_hal/common/mss_clint.o: in function `SysTick_Config':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:50:(.text.SysTick_Config+0x8): relocation truncated to fit: R_RISCV_HI20 against `.LC0'
./src/platform/mpfs_hal/common/mss_clint.o: in function `handle_m_timer_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:99:(.text.handle_m_timer_interrupt+0x32): relocation truncated to fit: R_RISCV_HI20 against `.L13'
./src/platform/mpfs_hal/common/mss_clint.o: in function `handle_m_soft_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:139:(.text.handle_m_soft_interrupt+0x26): relocation truncated to fit: R_RISCV_HI20 against `.L24'
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_DisableIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:852:(.text.PLIC_DisableIRQ+0x2e): relocation truncated to fit: R_RISCV_HI20 against `.L4'
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_ClaimIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:925:(.text.PLIC_ClaimIRQ+0x1a): relocation truncated to fit: R_RISCV_HI20 against symbol `plic_hart_lookup' defined in .rodata.plic_hart_lookup section in ./src/platform/mpfs_hal/common/mss_plic.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_CompleteIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:938:(.text.PLIC_CompleteIRQ+0x32): relocation truncated to fit: R_RISCV_HI20 against symbol `plic_hart_lookup' defined in .rodata.plic_hart_lookup section in ./src/platform/mpfs_hal/common/mss_plic.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `handle_m_ext_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_mtrap.c:696:(.text.handle_m_ext_interrupt+0x28): relocation truncated to fit: R_RISCV_HI20 against symbol `ext_irq_handler_table' defined in .data.ext_irq_handler_table section in ./src/platform/mpfs_hal/common/mss_mtrap.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `handle_local_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_mtrap.c:719:(.text.handle_local_interrupt+0x28): additional relocation overflows omitted from the output
collect2.exe: error: ld returned 1 exit status
make[1]: *** [makefile:79: fledgling_app_sw.elf] Error 1
make: *** [makefile:70: all] Error 2
"make all" terminated with exit code 2. Build might be incomplete.

11:44:47 Build Failed. 2 errors, 32 warnings. (took 41s.213ms)

Thanks, Geoff...

**Comment by geofflloyd** [comment]
I just tried starting with the stock "mpfs-lim.ld" linker file that builds fine and changed the following:
=> changed all the l2lim assignments to ddr_cached_32bit assignments
=> left the .l2_scratchpad and .ram_code sections as is since linker hopefully not using those sections
=> Add two new variable definitions to get around error in linking (existing linker variables are __app_stack_bottom_h1$ and __app_stack_top_h1):
        PROVIDE(__app_stack_bottom = .);
        PROVIDE(__app_stack_top = .);

Still get the same linker errors. They are pasted below for reference.
Can you explain what is going wrong?


./src/application/filesystem.o ./src/application/fw.o ./src/application/i2c_interface.o ./src/application/menu_main.o ./src/application/menu_rf_test.o ./src/application/modulation_out.o ./src/application/mpfs.o ./src/application/ring_buffer.o ./src/application/serial_console.o ./src/application/serial_port.o ./src/application/spi_interface.o ./src/application/systick.o ./src/application/usb_host_msc.o   
./src/platform/mpfs_hal/startup_gcc/newlib_stubs.o: in function `_sbrk':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/newlib_stubs.c:278:(.text._sbrk+0xc): relocation truncated to fit: R_RISCV_HI20 against symbol `__heap_end' defined in .heap section in fledgling_app_sw.elf
./src/platform/mpfs_hal/startup_gcc/system_startup.o: in function `u54_single_hart':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/system_startup.c:230:(.text.u54_single_hart+0x2c): relocation truncated to fit: R_RISCV_HI20 against `.L7'
./src/platform/mpfs_hal/startup_gcc/system_startup.o: in function `init_memory':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/startup_gcc/system_startup.c:487:(.text.init_memory+0x8): relocation truncated to fit: R_RISCV_HI20 against symbol `__text_end' defined in .text section in fledgling_app_sw.elf
./src/platform/mpfs_hal/common/mss_clint.o: in function `SysTick_Config':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:50:(.text.SysTick_Config+0x8): relocation truncated to fit: R_RISCV_HI20 against `.LC0'
./src/platform/mpfs_hal/common/mss_clint.o: in function `handle_m_timer_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:99:(.text.handle_m_timer_interrupt+0x32): relocation truncated to fit: R_RISCV_HI20 against `.L13'
./src/platform/mpfs_hal/common/mss_clint.o: in function `handle_m_soft_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_clint.c:139:(.text.handle_m_soft_interrupt+0x26): relocation truncated to fit: R_RISCV_HI20 against `.L24'
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_DisableIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:852:(.text.PLIC_DisableIRQ+0x2e): relocation truncated to fit: R_RISCV_HI20 against `.L4'
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_ClaimIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:925:(.text.PLIC_ClaimIRQ+0x1a): relocation truncated to fit: R_RISCV_HI20 against symbol `plic_hart_lookup' defined in .rodata.plic_hart_lookup section in ./src/platform/mpfs_hal/common/mss_plic.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `PLIC_CompleteIRQ':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\src\platform/mpfs_hal/common/mss_plic.h:938:(.text.PLIC_CompleteIRQ+0x32): relocation truncated to fit: R_RISCV_HI20 against symbol `plic_hart_lookup' defined in .rodata.plic_hart_lookup section in ./src/platform/mpfs_hal/common/mss_plic.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `handle_m_ext_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_mtrap.c:696:(.text.handle_m_ext_interrupt+0x28): relocation truncated to fit: R_RISCV_HI20 against symbol `ext_irq_handler_table' defined in .data.ext_irq_handler_table section in ./src/platform/mpfs_hal/common/mss_mtrap.o
./src/platform/mpfs_hal/common/mss_mtrap.o: in function `handle_local_interrupt':
C:\projects\sw9097\trunk\sw\fledgling_app_sw\Debug/../src/platform/mpfs_hal/common/mss_mtrap.c:719:(.text.handle_local_interrupt+0x28): additional relocation overflows omitted from the output
collect2.exe: error: ld returned 1 exit status
make[1]: *** [makefile:79: fledgling_app_sw.elf] Error 1
make: *** [makefile:70: all] Error 2
"make all" terminated with exit code 2. Build might be incomplete.

12:00:16 Build Failed. 2 errors, 0 warnings. (took 2s.711ms)

Thanks, Geoff...

**Reply by geofflloyd** [reply]
Can someone please point me to or send me the latest MPFS SW “platform”? 
I see that it has not be updated on the MPFS Git Repository since March 14th (with latest release tag Nov. 12, 2021).
https://github.com/polarfire-soc/platform
I assume that Microchip has internal release candidate versions of the SW “platform”.
Is there a scheduled date when the next Microchip SW “platform” release will occur?
After a lot more experimentation, I am guessing my problem has to do with some of the “platform” files.
I am currently using the Nov. 12, 2021 release tag of the SW "Platform" in my Application SW.
Thanks, Geoff...

**Comment by nitindeshpande** [comment]
@geofflloyd,
Sorry about the troubles, we will work with you through the support channel to debug the issues get you going with your project.
The next platform release is planned for End of July 2022.

**Comment by nitindeshpande** [comment, marked as answer]
@geofflloyd 
Below is the summary of the debug after which the combination of your custom bare metal bootloader and target application executing from DDR worked fine.
1) The custom bootloader did not wake up the target application processor (in this case U54_1) by raising a software interrupt to it. (Bootloader startup code running on E51 keeps it in WFI)
2) Bootloader jumpt-to-application() function was called from E51 to jump to DDR. The E51 should first wake-up the application processor (in this case U54_1) and U4_1  should then make a call to jumpt-to-application() so that it then jumps to DDR address.
3) The linker error _relocation truncated to fit: R_RISCV_HI20 against .L7'_ was because of the Code model selection as _medlow_ instead of _medany_
![image](https://user-images.githubusercontent.com/16042586/179400368-c72fe457-bc0f-4e3c-9461-5a97ba9b5983.png)


[value="ilg.gnumcueclipse.managedbuild.cross.riscv.option.target.codemodel.any"](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/acdddf3a903aa7e7b82961559965466760fa74fd/driver-examples/mss/mpfs-hal/mpfs-hal-run-from-ddr-u54-1/.cproject#L92)

Here are further readings on the topic of code models and RISC-V compiler options
https://www.sifive.com/blog/all-aboard-part-4-risc-v-code-models
https://gcc.gnu.org/onlinedocs/gcc/RISC-V-Options.html

Feedback by @geofflloyd
1) The build configurations described in [README.md](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/README.md) are useful but can be improved. 
     - We will add more snapshots and description of the specific SoftConsole project settings.
3) Better description of MPFS_HAL_LAST_HART and IMAGE_LOADED_BY_BOOTLOADER [here] could have avoided confusion
     -  Bare metal project designed as [simple bootloader (executing from eNVM)](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mpfs-hal/mpfs-hal-ddr-demo) or the [application executing from DDR.](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mpfs-hal/mpfs-hal-run-from-ddr-u54-1) are provided as starting point. We will describe the configuration parameters in more details for both these specific cases.
4) HSS Documentation did not provide much information on how to use it for bare metal applications.
     - Usage and documentation of HSS as bootloader for bare metal applications is updated in recent past; a [complete example](https://github.com/polarfire-soc/hart-software-services/tree/master/tools/hss-payload-generator#complete-example) is described here.
